diff -up src/plugins/preauth/pkinit/pkinit_crypto_openssl.c.libressl src/plugins/preauth/pkinit/pkinit_crypto_openssl.c
--- src/plugins/preauth/pkinit/pkinit_crypto_openssl.c.libressl	2019-08-08 11:48:05.738416757 +0200
+++ src/plugins/preauth/pkinit/pkinit_crypto_openssl.c	2019-08-08 15:44:47.186480414 +0200
@@ -187,7 +187,7 @@ pkinit_pkcs11_code_to_text(int err);
     (*_x509_pp) = PKCS7_cert_from_signer_info(_p7,_si)
 #endif
 
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
 
 /* 1.1 standardizes constructor and destructor names, renaming
  * EVP_MD_CTX_{create,destroy} and deprecating ASN1_STRING_data. */
@@ -3049,6 +3049,24 @@ cleanup:
 
 #if OPENSSL_VERSION_NUMBER >= 0x10100000L
 
+#if defined(LIBRESSL_VERSION_NUMBER)
+# define static_ASN1_SEQUENCE_END_name(stname, tname) \
+          ;\
+          static_ASN1_ITEM_start(tname) \
+                  ASN1_ITYPE_SEQUENCE,\
+                  V_ASN1_SEQUENCE,\
+                  tname##_seq_tt,\
+                  sizeof(tname##_seq_tt) / sizeof(ASN1_TEMPLATE),\
+                  NULL,\
+                  sizeof(stname),\
+                  #stname \
+          ASN1_ITEM_end(tname)
+
+# define static_ASN1_ITEM_start(itname) \
+          static ASN1_ITEM_start(itname)
+  
+#endif
+
 /*
  * We need to decode DomainParameters from RFC 3279 section 2.3.3.  We would
  * like to just call d2i_DHxparams(), but Microsoft's implementation may omit
