diff -up kernel-source/drivers/usb/serial/option.c.USB_option_reduce_interrupt-urb_logging_verbosity kernel-source/drivers/usb/serial/option.c
--- kernel-source/drivers/usb/serial/option.c.USB_option_reduce_interrupt-urb_logging_verbosity	2016-02-01 09:50:31.664871845 +0100
+++ kernel-source/drivers/usb/serial/option.c	2016-02-01 09:51:53.070525121 +0100
@@ -1917,8 +1917,11 @@ static void option_instat_callback(struc
 			dev_dbg(dev, "%s: type %x req %x\n", __func__,
 				req_pkt->bRequestType, req_pkt->bRequest);
 		}
-	} else
+	} else if (status == -ENOENT || status == -ESHUTDOWN) {
+		dev_dbg(dev, "%s: urb stopped: %d\n", __func__, status);
+	} else {
 		dev_err(dev, "%s: error %d\n", __func__, status);
+	}
 
 	/* Resubmit urb so we continue receiving IRQ data */
 	if (status != -ESHUTDOWN && status != -ENOENT) {
