From 9915494fdc0e499dd1a09faf01a3bec9df51d0dc Mon Sep 17 00:00:00 2001
From: Gernot Tenchio <gernot.tenchio@securepoint.de>
Date: Tue, 18 Oct 2016 11:40:34 +0200
Subject: [PATCH 3/3] configure.ac: check utmp harder

---
 configure.ac | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/configure.ac b/configure.ac
index 90bd278..735e2f2 100644
--- a/configure.ac
+++ b/configure.ac
@@ -522,7 +522,16 @@ AC_CHECK_HEADERS(sys/swap.h vm/anon.h, [], [have_sys_swap_h="no"],
 # For load module
 # For the processes plugin
 # For users module
-AC_CHECK_HEADERS(sys/loadavg.h linux/config.h utmp.h utmpx.h)
+AC_CHECK_HEADERS(sys/loadavg.h linux/config.h)
+AC_CHECK_FUNCS([getutent],
+  dnl Make sure getutent is supported by the C library
+  AC_CHECK_HEADERS([utmp.h])
+)
+AC_CHECK_FUNCS([getutxent],
+  dnl Make sure getutxent is supported by the C library
+  AC_CHECK_HEADERS([utmpx.h])
+)
+
 
 # For interface plugin
 AC_CHECK_HEADERS(ifaddrs.h)
-- 
2.7.4

