diff --git a/setconf.c b/setconf.c
index bfd0a3a..797c675 100644
--- a/setconf.c
+++ b/setconf.c
@@ -64,6 +64,14 @@ static bool sync_conf(struct wgdevice *file)
 
 	for_each_wgpeer(file, peer) {
 		pubkeys[i].pubkey = peer->public_key;
+		if (!(peer->flags & WGPEER_HAS_PRESHARED_KEY)) {
+			memset(peer->preshared_key, 0, WG_KEY_LEN);
+			peer->flags |= WGPEER_HAS_PRESHARED_KEY;
+		}
+		if (!(peer->flags & WGPEER_HAS_PERSISTENT_KEEPALIVE_INTERVAL)) {
+			peer->persistent_keepalive_interval = 0;
+			peer->flags |= WGPEER_HAS_PERSISTENT_KEEPALIVE_INTERVAL;
+		}
 		pubkeys[i].from_file = true;
 		++i;
 	}
