diff --git a/plugins/gssapi.c b/plugins/gssapi.c
index ff663da..fabb590 100644
--- a/plugins/gssapi.c
+++ b/plugins/gssapi.c
@@ -1436,9 +1436,6 @@ gssapi_server_mech_step(void *conn_context,
 
     if (text == NULL) return SASL_BADPROT;
 
-    params->utils->log(params->utils->conn, SASL_LOG_DEBUG,
-		       "GSSAPI server step %d\n", text->state);
-
     switch (text->state) {
 
     case SASL_GSSAPI_STATE_AUTHNEG:
@@ -1668,9 +1665,6 @@ static int gssapi_client_mech_step(void *conn_context,
     if (clientoutlen)
         *clientoutlen = 0;
     
-    params->utils->log(params->utils->conn, SASL_LOG_DEBUG,
-		       "GSSAPI client step %d", text->state);
-
     switch (text->state) {
 
     case SASL_GSSAPI_STATE_AUTHNEG:
