From b467cadb199c8884440fc40c861d0666831f2171 Mon Sep 17 00:00:00 2001
From: Gernot Tenchio <gernot.tenchio@securepoint.de>
Date: Sat, 19 Mar 2016 04:57:55 +0100

---
 clamd/session.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/clamd/session.c b/clamd/session.c
index b64dc77..d9798ad 100644
--- a/clamd/session.c
+++ b/clamd/session.c
@@ -64,7 +64,7 @@
 #include "session.h"
 #include "thrmgr.h"
 
-#ifndef HAVE_FDPASSING
+#ifndef HAVE_FD_PASSING
 #define FEATURE_FDPASSING 0
 #else
 #define FEATURE_FDPASSING 1
