diff --git a/src/read_config_lex.l b/src/read_config_lex.l
index b0d9e61..5c0656c 100644
--- a/src/read_config_lex.l
+++ b/src/read_config_lex.l
@@ -49,6 +49,7 @@ ip6_form1	{ip6_part}{0,7}"::"{ip6_part}{0,7}
 ip6_form2	({hex_255}":"){0,7}{hex_255}
 ip6		{ip6_form1}{ip6_cidr}?|{ip6_form2}{ip6_cidr}?
 string		[a-zA-Z][a-zA-Z0-9\.\-\_]*
+qstring		\"[a-zA-Z][a-zA-Z0-9\.\-\_]*\"
 persistent	[P|p][E|e][R|r][S|s][I|i][S|s][T|t][E|e][N|n][T|T]
 nack		[N|n][A|a][C|c][K|k]
 alarm		[A|a][L|l][A|a][R|r][M|m]
@@ -157,6 +158,7 @@ notrack		[N|n][O|o][T|t][R|r][A|a][C|c][K|k]
 			  return T_FTFW; }
 {notrack}		{ return T_NOTRACK; }
 {string}		{ yylval.string = strdup(yytext); return T_STRING; }
+{qstring}   { yylval.string = strdup(yytext + 1); yylval.string[strlen(yylval.string) - 1] = '\0'; return T_STRING;}
 
 {comment}	;
 {ws}		;
