diff -up conntrack-tools-1.4.0/configure.ac.ipv6_hack conntrack-tools-1.4.0/configure.ac
--- conntrack-tools-1.4.0/configure.ac.ipv6_hack	2017-03-21 15:41:00.867431009 +0100
+++ conntrack-tools-1.4.0/configure.ac	2017-03-21 15:42:19.696179471 +0100
@@ -73,31 +73,9 @@ dnl check for inet_pton
 AC_CHECK_FUNCS(inet_pton)
 dnl Some systems have it, but not IPv6
 if test "$ac_cv_func_inet_pton" = "yes" ; then
-AC_MSG_CHECKING(if inet_pton supports IPv6)
-AC_RUN_IFELSE([AC_LANG_SOURCE([[
-#ifdef HAVE_SYS_TYPES_H
-#include <sys/types.h>
-#endif
-#ifdef HAVE_SYS_SOCKET_H
-#include <sys/socket.h>
-#endif
-#ifdef HAVE_NETINET_IN_H
-#include <netinet/in.h>
-#endif
-#ifdef HAVE_ARPA_INET_H
-#include <arpa/inet.h>
-#endif
-int main()
-  {
-     struct in6_addr addr6;
-     if (inet_pton(AF_INET6, "::1", &addr6) < 1)
-        exit(1);
-     else
-        exit(0);
-  }
-  ]])],[ AC_MSG_RESULT(yes)
-       AC_DEFINE_UNQUOTED(HAVE_INET_PTON_IPV6, 1, [Define to 1 if inet_pton supports IPv6.])
-  ],[AC_MSG_RESULT(no)],[AC_MSG_RESULT(no)])
+cat >>confdefs.h <<_ACEOF
+#define HAVE_INET_PTON_IPV6 1
+_ACEOF
 fi
 
 # Checks for header files.
