diff --git a/bin/named/main.c b/bin/named/main.c
index 33421faeaa..26919aeeea 100644
--- a/bin/named/main.c
+++ b/bin/named/main.c
@@ -1029,11 +1029,11 @@ setup(void) {
 	isc_log_write(named_g_lctx, NAMED_LOGCATEGORY_GENERAL,
 		      NAMED_LOGMODULE_MAIN, ISC_LOG_NOTICE, "running on %s",
 		      named_os_uname());
-
+  /*
 	isc_log_write(named_g_lctx, NAMED_LOGCATEGORY_GENERAL,
 		      NAMED_LOGMODULE_MAIN, ISC_LOG_NOTICE, "built with %s",
 		      PACKAGE_CONFIGARGS);
-
+  */
 	isc_log_write(named_g_lctx, NAMED_LOGCATEGORY_GENERAL,
 		      NAMED_LOGMODULE_MAIN, ISC_LOG_NOTICE,
 		      "running as: %s%s%s", program_name, saved_command_line,
