From 4c5841f6db39ed36fc3a5fe4bf9f1e0d84b8fac7 Mon Sep 17 00:00:00 2001
From: Gernot Tenchio <gernot.tenchio@securepoint.de>
Date: Tue, 6 Aug 2019 14:46:36 +0200
Subject: [PATCH] auth/ntlm: unescape returned username

---
 src/auth/ntlm/UserRequest.cc | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/auth/ntlm/UserRequest.cc b/src/auth/ntlm/UserRequest.cc
index dcac22c..7b295bc 100644
--- a/src/auth/ntlm/UserRequest.cc
+++ b/src/auth/ntlm/UserRequest.cc
@@ -25,6 +25,7 @@
 #include "HttpRequest.h"
 #include "MemBuf.h"
 #include "SquidTime.h"
+#include "rfc1738.h"
 
 Auth::Ntlm::UserRequest::UserRequest() :
     authserver(nullptr),
@@ -312,7 +313,7 @@ Auth::Ntlm::UserRequest::HandleReply(void *data, const Helper::Reply &reply)
 
     case Helper::Okay: {
         /* we're finished, release the helper */
-        const char *userLabel = reply.notes.findFirst("user");
+        char *userLabel = (char *)reply.notes.findFirst("user");
         if (!userLabel) {
             auth_user_request->user()->credentials(Auth::Failed);
             safe_free(lm_request->server_blob);
@@ -320,6 +321,7 @@ Auth::Ntlm::UserRequest::HandleReply(void *data, const Helper::Reply &reply)
             debugs(29, DBG_CRITICAL, "ERROR: NTLM Authentication helper returned no username. Result: " << reply);
             break;
         }
+        rfc1738_unescape(userLabel);
         auth_user_request->user()->username(userLabel);
         auth_user_request->setDenyMessage("Login successful");
         safe_free(lm_request->server_blob);
