diff --git a/src/errorpage.cc b/src/errorpage.cc
index 5c1a8ea..a7cad7e 100644
--- a/src/errorpage.cc
+++ b/src/errorpage.cc
@@ -930,6 +930,9 @@ ErrorState::Convert(char token, bool building_deny_info_url, bool allowRecursion
             p = "[unknown method]";
         break;
 
+    case 'O':
+        do_quote = 0;
+        no_urlescape = 1;
     case 'o':
         p = request ? request->extacl_message.termedBuf() : external_acl_message;
         if (!p && !building_deny_info_url)
