From 76f79880de8678138c7c3e5856e1a8b616c45ea4 Mon Sep 17 00:00:00 2001
From: Gernot Tenchio <gernot.tenchio@securepoint.de>
Date: Mon, 20 Jan 2014 12:41:25 +0100

---
 src/ssl/gadgets.cc | 1 -
 1 file changed, 1 deletion(-)

diff --git a/src/ssl/gadgets.cc b/src/ssl/gadgets.cc
index 2688513..148564c 100644
--- a/src/ssl/gadgets.cc
+++ b/src/ssl/gadgets.cc
@@ -357,7 +357,6 @@ static int
 mimicExtensions(Security::CertPointer & cert, Security::CertPointer const &mimicCert, Security::CertPointer const &issuerCert)
 {
     static int extensions[]= {
-        NID_key_usage,
         NID_ext_key_usage,
         NID_basic_constraints,
         0
