diff -up rrdtool-1.4.7/configure.ac.noxml rrdtool-1.4.7/configure.ac
--- rrdtool-1.4.7/configure.ac.noxml	2013-11-05 22:09:47.021717194 +0100
+++ rrdtool-1.4.7/configure.ac	2013-11-05 22:33:45.845735415 +0100
@@ -111,7 +111,15 @@ if test $enable_rrd_graph != no; then
  AC_DEFINE([HAVE_RRD_GRAPH], [], [is rrd_graph supported by this install])
 fi
 
+AC_ARG_ENABLE(rrd_xml,AS_HELP_STRING([--disable-rrd_xml],[disable building of rrddump/rrdrestore]),
+[enable_rrd_xml=no],[enable_rrd_xml=yes])
+
+if test $enable_rrd_xml != no; then
+ AC_DEFINE([HAVE_RRD_XML], [], [is rrd_xml supported by this install])
+fi
+
 dnl Check if we run on a system that has fonts
+
 AC_ARG_WITH(rrd-default-font,
 [  --with-rrd-default-font=[OPTIONS]  set the full path to your default font.],
 [RRD_DEFAULT_FONT=$withval],[
@@ -532,6 +540,8 @@ AM_CONDITIONAL(BUILD_LIBWRAP,[test $have
 
 AM_CONDITIONAL(BUILD_RRDGRAPH,[test $enable_rrd_graph != no])
 
+AM_CONDITIONAL(BUILD_RRDXML,[test $enable_rrd_xml != no])
+
 CORE_LIBS="$LIBS"
 
 if test $enable_rrd_graph != no; then
@@ -540,7 +540,10 @@ EX_CHECK_ALL(cairo,      cairo_pdf_surfa
 EX_CHECK_ALL(cairo,      cairo_ps_surface_create,       cairo-ps.h,             cairo-ps,    1.10.2,  http://cairographics.org/releases/, "")
 EX_CHECK_ALL(pangocairo-1.0,  pango_cairo_context_set_font_options,  pango/pango.h,  pangocairo,  1.28.4,    http://ftp.gnome.org/pub/GNOME/sources/pango/1.28, "")
 fi
+
+if test $enable_rrd_xml != no; then
 EX_CHECK_ALL(xml2,       xmlParseFile,                  libxml/parser.h,        libxml-2.0,        2.7.8,  http://xmlsoft.org/downloads.html, "")
+fi
 
 if test "$EX_CHECK_ALL_ERR" = "YES"; then
   AC_MSG_ERROR([Please fix the library issues listed above and try again.])
@@ -964,6 +977,7 @@ echo
 echo "          With MMAP IO: $enable_mmap"
 echo "      Build rrd_getopt: $build_getopt"
 echo "       Build rrd_graph: $enable_rrd_graph"
+echo "         Build rrd_xml: $enable_rrd_xml"
 echo "       Static programs: $staticprogs"
 echo "          Perl Modules: $COMP_PERL"
 echo "           Perl Binary: $PERL"
diff -up rrdtool-1.4.7/src/Makefile.am.noxml rrdtool-1.4.7/src/Makefile.am
--- rrdtool-1.4.7/src/Makefile.am.noxml	2013-11-05 21:51:51.808567781 +0100
+++ rrdtool-1.4.7/src/Makefile.am	2013-11-05 21:53:15.245675107 +0100
@@ -38,13 +38,16 @@ RRD_C_FILES =		\
 	rrd_last.c	\
 	rrd_lastupdate.c	\
 	rrd_first.c	\
-	rrd_restore.c	\
-	rrd_dump.c	\
 	rrd_flushcached.c \
 	rrd_fetch.c	\
 	rrd_resize.c \
 	rrd_tune.c
 
+if BUILD_RRDXML
+RRD_C_FILES += rrd_restore.c   \
+	rrd_dump.c
+endif
+
 if BUILD_RRDGRAPH
 RRD_C_FILES += rrd_graph.c	\
 	rrd_graph_helper.c	\
diff -up rrdtool-1.4.7/src/rrd_tool.c.noxml rrdtool-1.4.7/src/rrd_tool.c
--- rrdtool-1.4.7/src/rrd_tool.c.noxml	2013-11-05 21:53:44.009097835 +0100
+++ rrdtool-1.4.7/src/rrd_tool.c	2013-11-05 22:07:00.477505536 +0100
@@ -645,7 +645,11 @@ int HandleInputLine(
     if (strcmp("create", argv[1]) == 0)
         rrd_create(argc - 1, &argv[1]);
     else if (strcmp("dump", argv[1]) == 0)
+#ifdef HAVE_RRD_XML
         rrd_dump(argc - 1, &argv[1]);
+#else
+        rrd_set_error("the instance of rrdtool has been compiled without dump/restore support");
+#endif
     else if (strcmp("info", argv[1]) == 0 || strcmp("updatev", argv[1]) == 0) {
         rrd_info_t *data;
 
@@ -666,7 +670,11 @@ int HandleInputLine(
                "  Copyright by Tobi Oetiker, 1997-2008 (%f)\n",
                rrd_version());
     else if (strcmp("restore", argv[1]) == 0)
+#ifdef HAVE_RRD_XML
         rrd_restore(argc - 1, &argv[1]);
+#else
+        rrd_set_error("the instance of rrdtool has been compiled without dump/restore support");
+#endif
     else if (strcmp("resize", argv[1]) == 0)
         rrd_resize(argc - 1, &argv[1]);
     else if (strcmp("last", argv[1]) == 0)
