diff --git a/fetchmail.h b/fetchmail.h
index b00ec7c5..f6ccb4d0 100644
--- a/fetchmail.h
+++ b/fetchmail.h
@@ -115,7 +115,7 @@ struct addrinfo;
 #define		MSGBUFSIZE	8192
 
 #define		NAMELEN		64	/* max username length */
-#define		PASSWORDLEN	256	/* max password length */
+#define		PASSWORDLEN	4096	/* max password length */
 #define		DIGESTLEN	33	/* length of MD5 digest */
 
 /* exit code values */
