diff -up sendmail-8.14.9/include/sm/bdb.h.fix-sloppy-DB-check sendmail-8.14.9/include/sm/bdb.h
--- sendmail-8.14.9/include/sm/bdb.h.fix-sloppy-DB-check	2015-08-04 10:11:01.883842846 +0200
+++ sendmail-8.14.9/include/sm/bdb.h	2015-08-04 10:11:24.159958888 +0200
@@ -19,7 +19,7 @@
 #  define DB_VERSION_MAJOR 1
 # endif /* ! DB_VERSION_MAJOR */
 
-# if DB_VERSION_MAJOR >= 4 && DB_VERSION_MINOR >= 1
+# if DB_VERSION_MAJOR > 4 || (DB_VERSION_MAJOR == 4 && DB_VERSION_MINOR >= 1)
 
 #  define DBTXN	NULL ,
 
