diff --git a/src/openvpn/ssl_verify.c b/src/openvpn/ssl_verify.c
index 4f3b61d6..111e8e15 100644
--- a/src/openvpn/ssl_verify.c
+++ b/src/openvpn/ssl_verify.c
@@ -331,6 +331,20 @@ print_nsCertType(int type)
     }
 }
 
+static int list_cmp(char *list, char *cn)
+{
+  int ret = -1;
+  char *pos = strstr(list, cn);
+  int cn_len = strlen(cn);
+  if (pos) {
+    if ((pos == list || (pos > list && *(pos - 1) == ',')) && ( *(pos + cn_len) == ',' || *(pos + cn_len) == '\0'))
+    {
+      ret = 0;
+    }
+  }
+  return ret;
+}
+
 /*
  * Verify the peer's certificate fields.
  *
@@ -393,7 +407,7 @@ verify_peer_cert(const struct tls_options *opt, openvpn_x509_cert_t *peer_cert,
         if ( (opt->verify_x509_type == VERIFY_X509_SUBJECT_DN
               && strcmp(opt->verify_x509_name, subject) == 0)
              || (opt->verify_x509_type == VERIFY_X509_SUBJECT_RDN
-                 && strcmp(opt->verify_x509_name, common_name) == 0)
+                 && list_cmp(opt->verify_x509_name, common_name) == 0)
              || (opt->verify_x509_type == VERIFY_X509_SUBJECT_RDN_PREFIX
                  && strncmp(opt->verify_x509_name, common_name,
                             strlen(opt->verify_x509_name)) == 0) )
