diff -up git/src/ssl/support.cc.typecast git/src/ssl/support.cc
--- git/src/ssl/support.cc.typecast	2022-10-13 09:57:24.847949363 +0200
+++ git/src/ssl/support.cc	2022-10-13 09:59:17.076182581 +0200
@@ -561,11 +561,11 @@ Ssl::VerifyCallbackParameters::At(Securi
 // "dup" function for SSL_get_ex_new_index("cert_err_check")
 #if SQUID_USE_CONST_CRYPTO_EX_DATA_DUP
 static int
-ssl_dupAclChecklist(CRYPTO_EX_DATA *, const CRYPTO_EX_DATA *, void *,
+ssl_dupAclChecklist(CRYPTO_EX_DATA *, const CRYPTO_EX_DATA *, void **,
                     int, long, void *)
 #else
 static int
-ssl_dupAclChecklist(CRYPTO_EX_DATA *, CRYPTO_EX_DATA *, void *,
+ssl_dupAclChecklist(CRYPTO_EX_DATA *, CRYPTO_EX_DATA *, void **,
                     int, long, void *)
 #endif
 {
