--- a/src/auth/auth-request.c.orig	2015-08-27 12:43:49.260353162 +0200
+++ b/src/auth/auth-request.c	2015-08-27 12:45:49.774405537 +0200
@@ -627,28 +627,28 @@
 	       auth_request_want_skip_passdb(request, next_passdb))
 		next_passdb = next_passdb->next;
 
-	if (*result == PASSDB_RESULT_OK) {
-		/* this passdb lookup succeeded, preserve its extra fields */
-		auth_fields_snapshot(request->extra_fields);
-		request->snapshot_have_userdb_prefetch_set =
-			request->userdb_prefetch_set;
-		if (request->userdb_reply != NULL)
-			auth_fields_snapshot(request->userdb_reply);
-	} else {
-		/* this passdb lookup failed, remove any extra fields it set */
-		auth_fields_rollback(request->extra_fields);
-		if (request->userdb_reply != NULL) {
-			auth_fields_rollback(request->userdb_reply);
-			request->userdb_prefetch_set =
-				request->snapshot_have_userdb_prefetch_set;
-		}
-	}
-
 	if (passdb_continue && next_passdb != NULL) {
 		/* try next passdb. */
                 request->passdb = next_passdb;
 		request->passdb_password = NULL;
 
+		if (*result == PASSDB_RESULT_OK) {
+			/* this passdb lookup succeeded, preserve its extra fields */
+			auth_fields_snapshot(request->extra_fields);
+			request->snapshot_have_userdb_prefetch_set =
+				request->userdb_prefetch_set;
+			if (request->userdb_reply != NULL)
+				auth_fields_snapshot(request->userdb_reply);
+		} else {
+			/* this passdb lookup failed, remove any extra fields it set */
+			auth_fields_rollback(request->extra_fields);
+			if (request->userdb_reply != NULL) {
+				auth_fields_rollback(request->userdb_reply);
+				request->userdb_prefetch_set =
+					request->snapshot_have_userdb_prefetch_set;
+			}
+		}
+
 		if (*result == PASSDB_RESULT_USER_UNKNOWN) {
 			/* remember that we did at least one successful
 			   passdb lookup */

