diff -rupN dovecot-2.2.14.orig/src/doveadm/doveadm-mail-fetch.c dovecot-2.2.14/src/doveadm/doveadm-mail-fetch.c
--- dovecot-2.2.14.orig/src/doveadm/doveadm-mail-fetch.c	2014-09-10 12:38:58.000000000 +0200
+++ dovecot-2.2.14/src/doveadm/doveadm-mail-fetch.c	2015-06-19 18:14:51.974858111 +0200
@@ -19,6 +19,9 @@
 #include "doveadm-mail.h"
 #include "doveadm-mailbox-list-iter.h"
 #include "doveadm-mail-iter.h"
+#include "maildir-storage.h"
+#include "maildir-uidlist.h"
+#include "maildir-sync.h"
 
 #include <stdio.h>
 
@@ -411,6 +414,31 @@ static int fetch_pop3_order(struct fetch
 	return 0;
 }
 
+static int fetch_fname(struct fetch_cmd_context *ctx)
+{
+	int maildir_file_do_callback_print(struct maildir_mailbox *mbox __attribute__((unused)),
+			const char *path, void *context __attribute__((unused)))
+	{
+		doveadm_print(path);
+
+		return 1;
+	}
+
+	int ret = 0;
+
+	if (strcmp(ctx->mail->box->storage->name, MAILDIR_STORAGE_NAME) == 0) {
+		struct maildir_mailbox *mbox = NULL;
+
+		mbox = (struct maildir_mailbox *) ctx->mail->box;
+		maildir_sync_refresh_flags_view(mbox);
+		ret = maildir_file_do(mbox, ctx->mail->uid, maildir_file_do_callback_print, NULL);
+
+		ret = ret < 0 ? -1 : 0;
+	}
+
+	return ret;
+}
+
 static const struct fetch_field fetch_fields[] = {
 	{ "user",          0,                        fetch_user },
 	{ "mailbox",       0,                        fetch_mailbox },
@@ -435,7 +463,8 @@ static const struct fetch_field fetch_fi
 	{ "imap.body",     MAIL_FETCH_IMAP_BODY,     fetch_imap_body },
 	{ "imap.bodystructure", MAIL_FETCH_IMAP_BODYSTRUCTURE, fetch_imap_bodystructure },
 	{ "pop3.uidl",     MAIL_FETCH_UIDL_BACKEND,  fetch_pop3_uidl },
-	{ "pop3.order",    MAIL_FETCH_POP3_ORDER,    fetch_pop3_order }
+	{ "pop3.order",    MAIL_FETCH_POP3_ORDER,    fetch_pop3_order },
+	{ "fname",         0,                        fetch_fname }
 };
 
 static const struct fetch_field *fetch_field_find(const char *name)
diff -rupN dovecot-2.2.14.orig/src/doveadm/Makefile.am dovecot-2.2.14/src/doveadm/Makefile.am
--- dovecot-2.2.14.orig/src/doveadm/Makefile.am	2014-09-16 14:23:48.000000000 +0200
+++ dovecot-2.2.14/src/doveadm/Makefile.am	2015-06-19 18:15:15.477859425 +0200
@@ -19,6 +19,8 @@ AM_CPPFLAGS = \
 	-I$(top_srcdir)/src/lib-imap \
 	-I$(top_srcdir)/src/lib-index \
 	-I$(top_srcdir)/src/lib-storage \
+	-I$(top_srcdir)/src/lib-storage/index \
+	-I$(top_srcdir)/src/lib-storage/index/maildir \
 	-I$(top_srcdir)/src/auth \
 	-DMODULEDIR=\""$(moduledir)"\" \
 	-DDOVEADM_MODULEDIR=\""$(doveadm_moduledir)"\" \
diff -rupN dovecot-2.2.14.orig/src/doveadm/Makefile.in dovecot-2.2.14/src/doveadm/Makefile.in
--- dovecot-2.2.14.orig/src/doveadm/Makefile.in	2014-10-14 16:17:06.000000000 +0200
+++ dovecot-2.2.14/src/doveadm/Makefile.in	2015-06-19 18:15:30.802860282 +0200
@@ -460,6 +460,8 @@ AM_CPPFLAGS = \
 	-I$(top_srcdir)/src/lib-imap \
 	-I$(top_srcdir)/src/lib-index \
 	-I$(top_srcdir)/src/lib-storage \
+	-I$(top_srcdir)/src/lib-storage/index \
+	-I$(top_srcdir)/src/lib-storage/index/maildir \
 	-I$(top_srcdir)/src/auth \
 	-DMODULEDIR=\""$(moduledir)"\" \
 	-DDOVEADM_MODULEDIR=\""$(doveadm_moduledir)"\" \
