diff -rupN dovecot-2.2.28.orig/src/doveadm/doveadm-mail-fetch.c dovecot-2.2.28/src/doveadm/doveadm-mail-fetch.c
--- dovecot-2.2.28.orig/src/doveadm/doveadm-mail-fetch.c	2017-03-03 11:43:51.631650396 +0100
+++ dovecot-2.2.28/src/doveadm/doveadm-mail-fetch.c	2017-03-03 11:52:29.525625612 +0100
@@ -15,6 +15,9 @@
 #include "mail-storage.h"
 #include "mail-search.h"
 #include "mail-namespace.h"
+#include "maildir-storage.h"
+#include "maildir-uidlist.h"
+#include "maildir-sync.h"
 #include "imap-msgpart.h"
 #include "doveadm-print.h"
 #include "doveadm-mail.h"
@@ -315,6 +318,31 @@ static int fetch_text_utf8(struct fetch_
 	return 0;
 }
 
+static int fetch_fname(struct fetch_cmd_context *ctx)
+{
+	int maildir_file_do_callback_print(struct maildir_mailbox *mbox __attribute__((unused)),
+			const char *path, void *context __attribute__((unused)))
+	{
+		doveadm_print(path);
+
+		return 1;
+	}
+
+	int ret = 0;
+
+	if (strcmp(ctx->mail->box->storage->name, MAILDIR_STORAGE_NAME) == 0) {
+		struct maildir_mailbox *mbox = NULL;
+
+		mbox = (struct maildir_mailbox *) ctx->mail->box;
+		maildir_sync_refresh_flags_view(mbox);
+		ret = maildir_file_do(mbox, ctx->mail->uid, maildir_file_do_callback_print, NULL);
+
+		ret = ret < 0 ? -1 : 0;
+	}
+
+	return ret;
+}
+
 static int fetch_size_physical(struct fetch_cmd_context *ctx)
 {
 	uoff_t size;
@@ -488,6 +516,7 @@ static const struct fetch_field fetch_fi
 	                   MAIL_FETCH_STREAM_BODY,   fetch_text },
 	{ "text.utf8",     MAIL_FETCH_STREAM_HEADER |
 	                   MAIL_FETCH_STREAM_BODY,   fetch_text_utf8 },
+	{ "fname",         0,                        fetch_fname },
 	{ "size.physical", MAIL_FETCH_PHYSICAL_SIZE, fetch_size_physical },
 	{ "size.virtual",  MAIL_FETCH_VIRTUAL_SIZE,  fetch_size_virtual },
 	{ "date.received", MAIL_FETCH_RECEIVED_DATE, fetch_date_received },
diff -rupN dovecot-2.2.28.orig/src/doveadm/Makefile.am dovecot-2.2.28/src/doveadm/Makefile.am
--- dovecot-2.2.28.orig/src/doveadm/Makefile.am	2017-03-03 11:43:51.631650396 +0100
+++ dovecot-2.2.28/src/doveadm/Makefile.am	2017-03-03 12:05:32.115588162 +0100
@@ -20,6 +20,8 @@ AM_CPPFLAGS = \
 	-I$(top_srcdir)/src/lib-imap \
 	-I$(top_srcdir)/src/lib-index \
 	-I$(top_srcdir)/src/lib-storage \
+	-I$(top_srcdir)/src/lib-storage/index \
+	-I$(top_srcdir)/src/lib-storage/index/maildir \
 	-I$(top_srcdir)/src/lib-imap-storage \
 	-I$(top_srcdir)/src/lib-http \
 	-I$(top_srcdir)/src/lib-dcrypt \
