diff -up syslog-ng-3.8.1/configure.ac.libdbi_no_host_path syslog-ng-3.8.1/configure.ac
--- syslog-ng-3.8.1/configure.ac.libdbi_no_host_path	2017-06-26 19:54:36.623158757 -0400
+++ syslog-ng-3.8.1/configure.ac	2017-06-26 19:54:56.424258604 -0400
@@ -749,7 +749,7 @@ if test "x$enable_sql" = "xyes" || test
 	PKG_CHECK_MODULES(LIBDBI, dbi >= $LIBDBI_MIN_VERSION, with_libdbi="1", with_libdbi="0")
 	if test "$with_libdbi" -eq 0; then
 		dnl if libdbi has no .pc file (e.g., Ubuntu Precise), try it without one
-		AC_CHECK_LIB(dbi, dbi_initialize_r, LIBDBI_LIBS="-ldbi"; LIBDBI_CFLAGS="-I/usr/include/dbi"; with_libdbi="1")
+		AC_CHECK_LIB(dbi, dbi_initialize_r, LIBDBI_LIBS="-ldbi"; LIBDBI_CFLAGS=""; with_libdbi="1")
 	fi
 
 	AC_MSG_CHECKING(whether to enable SQL support)
